cloudinit/sources/DataSourceVultr.py

```diff
@@ -0,0 +1,153 @@
+# Author: Eric Benner <ebenner@vultr.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+# Vultr Metadata API:
+# https://www.vultr.com/metadata/
+
+import json
+
+from cloudinit import log as log
+from cloudinit import sources
+from cloudinit import util
+
+import cloudinit.sources.helpers.vultr as vultr
+
+LOG = log.getLogger(__name__)
+BUILTIN_DS_CONFIG = {
+    'url': 'http://169.254.169.254',
+    'retries': 30,
+    'timeout': 2,
+    'wait': 2
+}
+
+
+class DataSourceVultr(sources.DataSource):
+
+    dsname = 'Vultr'
+
+    def __init__(self, sys_cfg, distro, paths):
+        super(DataSourceVultr, self).__init__(sys_cfg, distro, paths)
+        self.ds_cfg = util.mergemanydict([
+            util.get_cfg_by_path(sys_cfg, ["datasource", "Vultr"], {}),
+            BUILTIN_DS_CONFIG])
+        BUILTIN_DS_CONFIG['url'] = self.ds_cfg.get(
+            'url', BUILTIN_DS_CONFIG['url'])
+        BUILTIN_DS_CONFIG['retries'] = self.ds_cfg.get(
+            'retries', BUILTIN_DS_CONFIG['retries'])
+        BUILTIN_DS_CONFIG['timeout'] = self.ds_cfg.get(
+            'timeout', BUILTIN_DS_CONFIG['timeout'])
+        BUILTIN_DS_CONFIG['wait'] = self.ds_cfg.get(
+            'wait', BUILTIN_DS_CONFIG['wait'])
+
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
```

Comment:
**strongly suggest changing**

the less you do here (in `_get_data`) the easier testing is on your part.  the datasource is such a pain to instantiate in testing, so doing something like this is easier:

```
dsdata = get_datasource_data(vultr.generate_config(self.metadata())
self.metadata_full = dsdata['full']
self.metadata = dsdata['metadata']
self.userdata_raw = dsdata['user-data']
self.vendordata_raw = dsdata['vendor-data']
```

Then you can write `get_datasource_data()` in a way that is easy to  mock.  Its easy to test it with different metadata values, because you passed the result of the problematic 'get_metadata()' in.

In short, I'm saying make generate_config do more so that the result of calling it is really easy for the datasource to use.

Reply:
you did that too, thanks.